gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/auth/AbstractDelegationTokenBinding.java

```diff
@@ -0,0 +1,266 @@
+   * @param fs owning FS.
+   *
+   * @throws IOException failure.
+   */
+  public void bindToFileSystem(final URI uri,
+                               final GoogleHadoopFileSystemBase fs)
+      throws IOException {
+    Preconditions.checkState((canonicalUri == null),
+                             "bindToFileSystem called twice");
+    this.canonicalUri = requireNonNull(uri);
+    this.fileSystem = requireNonNull(fs);
+    this.service = new Text(uri.toString());
+    this.owner = fs.getOwner();
+  }
+
+
+  /**
+   * Create a delegation token for the user.
+   * This will only be called if a new DT is needed, that is: the
+   * filesystem has been deployed unbound.
+   *
+   * @return the token
+   *
+   * @throws IOException if one cannot be created
+   */
+  public Token<AbstractGCPTokenIdentifier> createDelegationToken(String renewer)
+      throws IOException {
+    Text renewerText = new Text();
+    if (renewer != null) {
+      renewerText.set(renewer);
+    }
+
+    AbstractGCPTokenIdentifier tokenIdentifier =
+        requireNonNull(createTokenIdentifier(renewerText),
+                       "Token identifier");
+
+    Token<AbstractGCPTokenIdentifier> token =
+        new Token<>(tokenIdentifier, secretManager);
+    token.setKind(getKind());
+    token.setService(service);
+    logger.atFine().log("Created token %s with token identifier %s",
+        token, tokenIdentifier);
+    return token;
+  }
+
+  /**
+   * Create a token identifier with all the information needed
+   * to be included in a delegation token.
+   * This is where session credentials need to be extracted, etc.
+   * This will only be called if a new DT is needed, that is: the
+   * filesystem has been deployed unbound.
+   *
+   * If {@link #createDelegationToken}
+   * is overridden, this method can be replaced with a stub.
+   *
+   * @return the token data to include in the token identifier.
+   *
+   * @throws IOException failure creating the token data.
+   */
+  public abstract AbstractGCPTokenIdentifier createTokenIdentifier(Text renewer)
+      throws IOException;
+
+
+  /**
+   * Create a token identifier with all the information needed
+   * to be included in a delegation token.
+   * This is where session credentials need to be extracted, etc.
+   * This will only be called if a new DT is needed, that is: the
+   * filesystem has been deployed unbound.
+   *
+   * If {@link #createDelegationToken}
+   * is overridden, this method can be replaced with a stub.
+   *
+   * @return the token data to include in the token identifier.
+   *
+   * @throws IOException failure creating the token data.
+   */
+  public abstract AbstractGCPTokenIdentifier createTokenIdentifier()
+      throws IOException;
+
+
+  /**
+   * Create a new subclass of {@link AbstractGCPTokenIdentifier}.
+   * This is used in the secret manager.
+   * @return an empty identifier.
+   */
+  public abstract AbstractGCPTokenIdentifier createEmptyIdentifier();
```

Comment:
Could you clarify exactly what this is? Does it return a dummy object, or does it have a specific purpose?

Reply:
It does indeed return an "empty" token identifier. It is used by at least the "dummy" SecretManager, since it doesn't need a real identifier, but it does need to satisfy the contract.